### PR TITLE
Interactivity API: Use Script Modules data filter

### PR DESCRIPTION
**Pull request by sirreal**
https://core.trac.wordpress.org/changeset/58579 introduced an API for passing data to Script Modules similar to what Interactivity API does.

Instead of its own implementation, change the Interactivity API to use the Core API via the `script_module_data_{$module_id}` filter.

Todo:
- [x] All of the print/encode tests can be removed (that testing is now part of #6682).
- [x] Add tests on the filter being used correctly.

✅  ~This will not work until client scripts are updated. The `@wordpress/interactivity` Script Module was updated in https://github.com/WordPress/gutenberg/pull/61658 to try to get client data from an updated element ID. This requires `@wordpress/interactivity@>5.7.0`.~

Trac ticket: https://core.trac.wordpress.org/ticket/61512

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-script-modules.php [hunk 1]
```diff
@@ -363,4 +366,119 @@ private function get_src( string $id ): string {
...
+			 *
+			 * If the filter returns no data (an empty array), nothing will be embedded in the page.
+			 *
+			 * The data for a given Script Module, if provided, will be JSON serialized in a script
+			 * tag with an ID of the form `wp-scriptmodule-data_{$module_id}`.
+			 *
+			 * The data can be read on the client with a pattern like this:
+			 *
+			 * @example
+			 *   const dataContainer = document.getElementById( 'wp-scriptmodule-data_MyScriptModuleID' );
```
Comment (gziolo) [inline src/wp-includes/class-wp-script-modules.php]: See the [comment in the proposal](https://make.wordpress.org/core/2024/05/06/proposal-server-to-client-data-sharing-for-script-modules/#comment-46628) that was touching on this particular example:

> I understand the benefits here, but would love to see a better initialization story. If I’m understanding correctly, any module that wants to access data will need some version of the following with only their data script handle being replaced:
> [the same code snippet]
> Is there any way we could simplify data access and abstract this to reduce the amount of boilerplate required for each module that wants to access server data?

Comment (sirreal) [reply src/wp-includes/class-wp-script-modules.php]: I'll move this review to https://github.com/WordPress/wordpress-develop/pull/6682.

This PR is changing Interactivity API to use that.

Comment (gziolo) [reply src/wp-includes/class-wp-script-modules.php]: Ah, sorry about that. I didn't realize that it is part of another PR.

**Comment by sirreal** [comment]
@luisherranz @cbravobernal @darerodz I'd love to get an Interactivity API review of this 🙏

**Review by luisherranz — APPROVED** [review]
Looks good to me 🙂

Good trick to update the [JS handle](https://github.com/WordPress/gutenberg/pull/62170) beforehand 😄

**Review by cbravobernal — APPROVED** [review]
LGTM.

**Comment by sirreal** [comment]
Suggested commit message:

```
Interactivity API: Use Script Modules filter for store and config data

A dedicated API exists for passing data to Script Modules implemented in
changeset [58579]. Use this Core API instead of a custom implementation
for Interactivity API to pass data to the client.

Developed in https://github.com/WordPress/wordpress-develop/pull/6683.

Props jonsurrell, gziolo, luisherranz, cbravobernal.
Fixes #61512.
```

**Comment by ockham** [comment]
Committed to Core in https://core.trac.wordpress.org/changeset/58729.

**Pull request closed without merging**